documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md

```diff
@@ -609,3 +617,52 @@ monitor the first production data refreshes closely.
 
 [^2]:
     https://towardsdatascience.com/using-apache-airflow-dockeroperator-with-docker-compose-57d0217c8219
+
+[^3]:
+    See #4456 for further context on this. The filtering is a _necessary_ step
+    of the data refresh we need to carry forward even after removing the other
+    cleanup steps. There are a number of ways we can accomplish this, but by far
+    the easiest would be to directly map the functionality of the ingestion
+    server on this step within a single Airflow task. The steps for this task
+    are as follows:
+
+    1. Get a batch of records from the database using `CLEANUP_BUFFER_SIZE`
+    2. Divide batch up into `multiprocessing.cpu_count()` subbatches
+    3. Split the filtering up into separate workers using multiprocessing
+    4. On each process
+       1. Create a new DB connection & cursor per worker
+       2. Iterate through each record
+          1. Remove tags below confidence level
+          2. Remove tags that need to be filtered (denylisted, machine-generated
+             filter list, provider, etc)
+          3. Only surface the record if it needs to be changed
+          4. Update each records one by one with a single `UPDATE`
+       3. Commit cursor and close connection
+    5. Repeat steps 1-4 until all batches are consumed
+
+    The Airflow scheduler container has access to 4 cores, which is the same as
+    the ingestion server where this step was originally running. At present it
+    takes about 8 hours for all cleanup steps, but that includes the URL
+    cleaning which is certainly more time intensive than the tag filtering since
+    it makes outbound requests. Running the tag filtering on Airflow should not
+    impact any of the other running tasks or saturate the instance.
```

Comment:
> but that includes the URL cleaning which is certainly more time intensive than the tag filtering since it makes outbound requests.

Great point I totally hadn't thought of!